pkg/componentsregistry/componentsregistry.go

```diff
@@ -32,25 +32,31 @@ type ComponentHandler interface {
 	UpdateDSCStatus(dsc *dscv1.DataScienceCluster, obj client.Object) error
 }
 
-var registry = []ComponentHandler{}
+// Registry is a struct that maintains a list of registered ComponentHandlers.
+type Registry struct {
+	handlers []ComponentHandler
+}
 
-// Add registers a new component handler
+// Add registers a new ComponentHandler to the registry.
 // not thread safe, supposed to be called during init.
-// TODO: check if init() can be called in parallel.
-func Add(ch ComponentHandler) {
-	registry = append(registry, ch)
+func (r *Registry) Add(ch ComponentHandler) {
+	r.handlers = append(r.handlers, ch)
 }
 
-// ForEach iterates over all registered component handlers
+// ForEach iterates over all registered ComponentHandlers and applies the given function.
+// If any handler returns an error, that error is collected and returned at the end.
 // With go1.23 probably https://go.dev/blog/range-functions can be used.
-func ForEach(f func(ch ComponentHandler) error) error {
+func (r *Registry) ForEach(f func(ch ComponentHandler) error) error {
 	var errs *multierror.Error
-	for _, ch := range registry {
+	for _, ch := range r.handlers {
 		errs = multierror.Append(errs, f(ch))
 	}
+
 	return errs.ErrorOrNil()
 }
 
+var R = &Registry{}
```

Comment:
@ykaliuta I moved the implementation to a struct so that I can use it for unit-testing the controller because it is not easy injecting fault behaviors in an existing component without impacting the code. 

Let me know if this is accetable

Reply:
if I got it right, it is just about splitting the registry refactoring to a dedicated PR/commit, right ?

Reply:
Not exactly. It keeps old api in the places where it's not needed. I described it in the commit message as usually.

Reply:
> Ah yes, sorry, got confused by the github diff

I reused your patch as much as possible :)

Reply:
rebased on #1502